illumos-utils/src/opte/non_illumos.rs

```diff
@@ -43,3 +70,268 @@ pub fn delete_all_xde_devices(log: &Logger) -> Result<(), Error> {
+pub(crate) struct State {
+    pub ports: HashMap<String, PortData>,
+    pub underlay_initialized: bool,
+}
+
+const NO_RESPONSE: NoResp = NoResp { unused: 99 };
+static OPTE_STATE: OnceLock<Mutex<State>> = OnceLock::new();
+
+fn opte_state() -> &'static Mutex<State> {
+    OPTE_STATE.get_or_init(|| {
+        Mutex::new(State { ports: HashMap::new(), underlay_initialized: true })
+    })
+}
+
+/// Simulated handle to OPTE
+pub struct Handle;
+
+impl Handle {
+    /// Create a new handle to the test OPTE kernel module.
+    pub fn new() -> Result<Self, OpteError> {
+        Ok(Self)
+    }
+
+    /// Helper to get the OPTE state.
+    #[cfg(test)]
+    pub(crate) fn state(&self) -> &'static Mutex<State> {
+        opte_state()
+    }
+
+    /// Add a new port.
+    pub fn create_xde(
+        &self,
+        name: &str,
+        cfg: VpcCfg,
+        _: DhcpCfg,
+        _: bool,
+    ) -> Result<NoResp, OpteError> {
+        let name = name.to_string();
+        let IpCfg::Ipv4(ip_cfg) = cfg.ip_cfg else {
+            unimplemented!("IPv6 support");
+        };
+        let ephemeral_ip4_addr =
+            ip_cfg.external_ips.snat.as_ref().map(|snat| snat.external_ip);
+        let port = PortInfo {
+            name: name.clone(),
+            mac_addr: cfg.guest_mac,
+            ip4_addr: Some(ip_cfg.private_ip),
+            ephemeral_ip4_addr,
+            floating_ip4_addrs: None,
+            ip6_addr: None,
+            ephemeral_ip6_addr: None,
+            floating_ip6_addrs: None,
+            state: "created".to_string(),
+        };
+        let mut state = opte_state().lock().unwrap();
+        anyhow::ensure!(
+            state.underlay_initialized,
+            "Underlay is not initialized"
+        );
+        assert!(
+            state
+                .ports
+                .insert(name, PortData { port, routes: Vec::new() })
+                .is_none(),
+            "duplicate OPTE ports"
+        );
+        Ok(NO_RESPONSE)
+    }
+
+    pub fn delete_xde(&self, name: &str) -> Result<NoResp, OpteError> {
+        let _ = opte_state().lock().unwrap().ports.remove(name);
+        Ok(NO_RESPONSE)
+    }
+
+    /// Set new firewall rules.
+    pub fn set_fw_rules(&self, _: &SetFwRulesReq) -> Result<NoResp, OpteError> {
+        Ok(NO_RESPONSE)
+    }
+
+    /// Add a new router entry to OPTE.
+    pub fn add_router_entry(
+        &self,
+        req: &AddRouterEntryReq,
+    ) -> Result<NoResp, OpteError> {
+        let mut inner = opte_state().lock().unwrap();
+        let Some(PortData { routes, .. }) = inner.ports.get_mut(&req.port_name)
+        else {
+            anyhow::bail!("No such port '{}'", req.port_name);
+        };
+        routes.push(req.into());
+        Ok(NO_RESPONSE)
+    }
+
+    /// Allow traffic to / from a CIDR block on a port.
+    pub fn allow_cidr(
+        &self,
+        _: &str,
+        _: IpCidr,
+        _: Direction,
+    ) -> Result<NoResp, OpteError> {
+        todo!()
```

Comment:
Maybe `unimplemented!("Not yet used in tests")` like the others?

Reply:
Done in 2024b245c